### PR TITLE
fix: fb pixel test case refactor

### PR DESCRIPTION
**Pull request by shrouti1507**
## What are the changes introduced in this PR?

Write a brief explainer on your code changes.

## What is the related Linear task?

Resolves INT-1472

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

<hr>

### Developer checklist

- [x] My code follows the style guidelines of this project

- [x] **No breaking changes are being introduced.**

- [x] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [x] Is the PR limited to one linear task?

- [x] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3075/test-report.html

File: test/integrations/destinations/facebook_pixel/processor/data1.ts [hunk 1]
```diff
@@ -0,0 +1,6463 @@
+import { VERSION } from '../../../../../src/v0/destinations/facebook_pixel/config';
```
Comment (shrouti1507) [inline test/integrations/destinations/facebook_pixel/processor/data1.ts:1]: this is the previous test cases, will remove this file before merge

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 2]
```diff
@@ -0,0 +1,483 @@
+import { VERSION } from '../../../../../src/v0/destinations/facebook_pixel/config';
+import { overrideDestination, generateTrackPayload } from '../../../testUtils';
+
+const commonDestination = {
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:4]: Use the `Destination` Type for this variable

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+const commonPropertiesWithoutProductArray = {
+  category: 'dummy',
+  quantity: 10,
+  value: 100,
+  product_id: '12345',
+};
+
+const commonTimestamp = new Date('2023-10-14');
+
+export const configLevelFeaturesTestData = [
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:83]:
```suggestion
export const configLevelFeaturesTestData : ProcessorTestData[] = [
```

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+          },
+        ],
+      },
+    },
+    output: {
+      response: {
+        status: 200,
+        body: [
+          {
+            output: {
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:113]: https://github.com/rudderlabs/rudder-transformer/blob/511cfacd3b887bf6c4f5f16b9e6b3ed9a4dc6bd8/test/integrations/destinations/klaviyo/processor/groupTestData.ts#L77

use this function to reduce boiler plate efforts

**Review by utsabc — CHANGES_REQUESTED** [review]

**Comment by utsabc** [comment]
Please add what are scenarios that you were able to add on the existing tests cases, the number of new scenarios added

File: src/v0/destinations/facebook_pixel/utils.js [hunk 5]
```diff
@@ -30,7 +30,7 @@ const formatRevenue = (revenue) => {
  */
 const getActionSource = (payload, channel) => {
   let actionSource = 'other';
-  if (payload.action_source) {
+  if (payload?.action_source) {
```
Comment (utsabc) [inline src/v0/destinations/facebook_pixel/utils.js:33]: Why are we having optional chaining for `payload`?

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+  product_id: '12345',
+};
+
+const commonTimestamp = new Date('2023-10-14');
+
+export const configLevelFeaturesTestData: ProcessorTestData[] = [
+  {
+    id: 'facebook_pixel-test-1',
+    name: 'facebook_pixel',
+    description: 'config feature : limitedDataUSage switched on',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:94]: Can we add some facebook docs link in the description for this feature or you can add more context by adding a `comment` field

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+  Enabled: true,
+};
+
+const commonUserTraits = {
+  email: 'abc@gmail.com',
+  anonymousId: 'c82cbdff-e5be-4009-ac78-cdeea09ab4b1',
+  event_id: '12345',
+};
+
+const piiPropertiesForAllowDeny = {
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:66]: can we add inline comments to describe what is the purpose of this?

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+            output: transformResultBuilder({
+              version: '1',
+              type: 'REST',
+              method: 'POST',
+              endpoint: `https://graph.facebook.com/${VERSION}/dummyPixelId/events?access_token=09876`,
+              headers: {},
+              params: {},
+              FORM: {
+                data: [
+                  '{"user_data":{"external_id":"3ffc8a075f330402d82aa0a86c596b0d2fe70df38b22c5be579f86a18e4aca47","em":"48ddb93f0b30c475423fe177832912c5bcdce3cc72872f8051627967ef278e08","client_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:84.0) Gecko/20100101 Firefox/84.0"},"event_name":"ViewContent","event_time":1697241600,"event_id":"12345","action_source":"website","data_processing_options":"val1","data_processing_options_country":"val2","data_processing_options_state":"val3","custom_data":{"category":"dummy","quantity":10,"value":100,"product_id":"12345","content_ids":["dummy"],"content_type":"product_group","contents":[{"id":"dummy","quantity":1}],"content_category":"dummy","currency":"USD"}}',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:133]: can we declare this a json then use JSON.strigify to make it a string if possible?

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+      },
+    },
+  },
+  {
+    id: 'facebook_pixel-test-2',
+    name: 'facebook_pixel',
+    scenario: 'configuration',
+    description:
+      'config feature : ContentCategoryMapping table is filled up, and category is passed with properties',
+    successCriteria: 'Response should contain category mapped to newClothing',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:152]: Need more context (Write test details as if other presion has less context about fb)

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+            output: transformResultBuilder({
+              version: '1',
+              type: 'REST',
+              method: 'POST',
+              endpoint: `https://graph.facebook.com/${VERSION}/dummyPixelId/events?access_token=09876`,
+              headers: {},
+              params: {},
+              FORM: {
+                data: [
+                  '{"user_data":{"external_id":"3ffc8a075f330402d82aa0a86c596b0d2fe70df38b22c5be579f86a18e4aca47","em":"48ddb93f0b30c475423fe177832912c5bcdce3cc72872f8051627967ef278e08","client_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:84.0) Gecko/20100101 Firefox/84.0"},"event_name":"ViewContent","event_time":1697241600,"event_id":"12345","action_source":"website","data_processing_options":"val1","data_processing_options_country":"val2","data_processing_options_state":"val3","custom_data":{"category":"clothing","quantity":10,"value":100,"product_id":"12345","content_ids":["clothing"],"content_type":"newClothing","contents":[{"id":"clothing","quantity":1}],"content_category":"clothing","currency":"USD"}}',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:189]: Same as above comment

File: test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts [hunk 11]
```diff
@@ -0,0 +1,705 @@
...
+  Enabled: true,
+};
+
+const commonUserTraits = {
+  email: 'abc@gmail.com',
+  anonymousId: 'c82cbdff-e5be-4009-ac78-cdeea09ab4b1',
+  event_id: '12345',
+};
+
+const commonUserProperties = {
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts:54]: This is not used

File: test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts [hunk 12]
```diff
@@ -0,0 +1,705 @@
...
+  errorType: 'instrumentation',
+  destType: 'FACEBOOK_PIXEL',
+  module: 'destination',
+  implementation: 'native',
+  feature: 'processor',
+};
+
+export const ecommTestData: ProcessorTestData[] = [
+  {
+    id: 'facebook_pixel-test-1',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts:107]: Make sure every test has unique Id, for all the fb_pixel test data

File: test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts [hunk 13]
```diff
@@ -0,0 +1,705 @@
...
+            output: transformResultBuilder({
+              version: '1',
+              type: 'REST',
+              method: 'POST',
+              endpoint: `https://graph.facebook.com/${VERSION}/dummyPixelId/events?access_token=09876`,
+              headers: {},
+              params: {},
+              FORM: {
+                data: [
+                  '{"user_data":{"external_id":"3ffc8a075f330402d82aa0a86c596b0d2fe70df38b22c5be579f86a18e4aca47","em":"48ddb93f0b30c475423fe177832912c5bcdce3cc72872f8051627967ef278e08","client_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:84.0) Gecko/20100101 Firefox/84.0"},"event_name":"ViewContent","event_time":1697241600,"event_id":"12345","action_source":"website","custom_data":{"category":"dummy","quantity":10,"value":100,"product_id":"12345","content_ids":["dummy"],"content_type":"product_group","contents":[{"id":"dummy","quantity":1}],"content_category":"dummy","currency":"USD"}}',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts:149]: Difficult to read this as string like I mentioned if we can use JSON.stringify

Comment (utsabc) [reply test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts:149]: On top of a json object

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts (hunk 2 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts]: added

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts (hunk 3 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts]: edited

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts (hunk 4 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts]: used

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts [hunk 14]
```diff
@@ -0,0 +1,577 @@
...
+    advancedMapping: false,
+    whitelistPiiProperties: [
+      {
+        whitelistPiiProperties: 'email',
+      },
+    ],
+  },
+  Enabled: true,
+};
+const commonMessage = {
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:52]: We could have use fxns for `testUtils`

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts [hunk 15]
```diff
@@ -0,0 +1,577 @@
...
+        body: [
+          {
+            message: generateSimplifiedPageOrScreenPayload(
+              {
+                event: 'TestEven001',
+                sentAt: '2021-01-25T16:12:02.048Z',
+                userId: 'sajal12',
+                context: {
+                  traits: {},
+                },
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:179]: Any reason for passing empty `traits`?

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:231]: same here

**Comment by utsabc** [comment]
Also in terms of coverage whats the coverage diff btw old test cases and new ones

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts (hunk 9 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts:152]: edited

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts (hunk 15 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:179]: no reason, just wanted to keep it short.

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts (hunk 16 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:231]: added

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts (hunk 6 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts]: updated

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts (hunk 8 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts]: done

File: test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts (hunk 11 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts]: removed

File: test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts (hunk 12 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts]: updated

File: test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts (hunk 13 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/ecommTestData.ts]: done

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts (hunk 15 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts]: updated with minimal values

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts (hunk 16 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts]: updated with minimal values

File: src/v0/destinations/am/utils.js [hunk 17]
```diff
@@ -123,7 +123,6 @@ const validateEventType = (evType) => {
   }
 };
 
-
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply src/v0/destinations/am/utils.js]: same

File: src/v0/destinations/am/utils.js [hunk 18]
```diff
@@ -187,5 +186,5 @@ module.exports = {
   getEventId,
   getUnsetObj,
   validateEventType,
-  userPropertiesPostProcess
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply src/v0/destinations/am/utils.js]: not sure from where it is coming, have not made any change there

File: test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts (hunk 10 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/configLevelFeaturesTestData.ts]: done

File: src/v0/destinations/facebook_pixel/utils.js (hunk 5 above)
Comment (utsabc) [reply src/v0/destinations/facebook_pixel/utils.js:33]: Please update as discussed

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts [hunk 19]
```diff
@@ -0,0 +1,630 @@
...
+            output: transformResultBuilder({
+              version: '1',
+              type: 'REST',
+              method: 'POST',
+              endpoint: `https://graph.facebook.com/${VERSION}/dummyPixelId/events?access_token=09876`,
+              headers: {},
+              params: {},
+              FORM: {
+                data: [
+                  '{"user_data":{"external_id":"5994471abb01112afcc18159f6cc74b4f511b99806da59b3caf5a9c173cacfc5","client_ip_address":"0.0.0.0","client_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"},"event_name":"Viewed page ApplicationLoaded","event_time":1697278611,"event_source_url":"jkl","event_id":"5e10d13a-bf9a-44bf-b884-43a9e591ea71","action_source":"website","custom_data":{"path":"/abc","referrer":"xyz","search":"def","title":"ghi","url":"jkl"}}',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:145]: Please update this

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts [hunk 20]
```diff
@@ -0,0 +1,630 @@
...
+            output: transformResultBuilder({
+              version: '1',
+              type: 'REST',
+              method: 'POST',
+              endpoint: `https://graph.facebook.com/${VERSION}/dummyPixelId/events?access_token=09876`,
+              headers: {},
+              params: {},
+              FORM: {
+                data: [
+                  '{"user_data":{"external_id":"470582f368e5aeec2cf487decd1e125b7d265e8b0b06b74a25e999e93bfb699f"},"event_name":"PageView","event_time":1697297576,"event_source_url":"jkl","action_source":"website","custom_data":{"path":"/abc","referrer":"xyz","search":"def","title":"ghi","url":"jkl"}}',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:351]: Please update

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts [hunk 21]
```diff
@@ -0,0 +1,630 @@
...
+            output: transformResultBuilder({
+              version: '1',
+              type: 'REST',
+              method: 'POST',
+              endpoint: `https://graph.facebook.com/${VERSION}/dummyPixelId/events?access_token=09876`,
+              headers: {},
+              params: {},
+              FORM: {
+                data: [
+                  '{"user_data":{"external_id":"5994471abb01112afcc18159f6cc74b4f511b99806da59b3caf5a9c173cacfc5","client_ip_address":"0.0.0.0","client_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36"},"event_name":"PageView","event_time":1697278611,"event_source_url":"jkl","event_id":"5e10d13a-bf9a-44bf-b884-43a9e591ea71","action_source":"website","custom_data":{"path":"/abc","referrer":"xyz","search":"def","title":"ghi","url":"jkl"}}',
```
Comment (utsabc) [inline test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:399]: Please update

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts (hunk 20 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts]: done

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts (hunk 21 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts]: done

**Review by sandeepdsvs — APPROVED** [review]
LGTM

File: test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts (hunk 14 above)
Comment (shrouti1507) [reply test/integrations/destinations/facebook_pixel/processor/pageScreenTestData.ts:52]: done

**Review by utsabc — APPROVED** [review]

**Pull request merged**